flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorStateChangelogStoragesManager.java

```diff
@@ -135,28 +148,109 @@ public void releaseStateChangelogStorageForJob(@Nonnull JobID jobId) {
         }
     }
 
+    @Nullable
+    StateChangelogStorageView<?> stateChangelogStorageViewForJob(
+            @Nonnull JobID jobID,
+            Configuration configuration,
+            ChangelogStateHandle changelogStateHandle)
+            throws IOException {
+        if (closed) {
+            throw new IllegalStateException(
+                    "TaskExecutorStateChangelogStoragesManager is already closed and cannot "
+                            + "register a new StateChangelogStorageView.");
+        }
+
+        if (!(changelogStateHandle instanceof ChangelogStateHandleStreamImpl)) {
+            return StateChangelogStorageLoader.loadFromStateHandle(
+                    configuration, changelogStateHandle);
+        }
+
+        synchronized (lock) {
+            Optional<StateChangelogStorageView<ChangelogStateHandleStreamImpl>> storageView =
+                    changelogStorageViewsByJobId.get(jobID);
+
+            if (storageView == null) {
+                StateChangelogStorageView<?> loaded =
+                        StateChangelogStorageLoader.loadFromStateHandle(
+                                configuration, changelogStateHandle);
+                changelogStorageViewsByJobId.put(
+                        jobID,
+                        Optional.of(
+                                (StateChangelogStorageView<ChangelogStateHandleStreamImpl>)
+                                        loaded));
```

Comment:
Could you explain why is it `Optional`?

Reply:
This should indeed always return a non-null object, I will amend it later.